### PR TITLE
feat: Implement the `StorageManager` and `InMemoryCache`

### PR DESCRIPTION
**Pull request by Reto911**
`StorageManager` and `InMemoryCache` are implemented in this PR, they are moved from #431.

### CODE REVIEW

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+        ino: INum,
+        block_id: usize,
+        block: Block,
+    ) -> Option<(BlockCoordinate, Block)>
+    where
+        P: EvictPolicy<BlockCoordinate>,
+    {
+        let mut empty_file_cache = None;
+        let to_be_evicted = self.policy.put(BlockCoordinate(ino, block_id));
+        let guard = pin();
```
Comment (xiaguan) [inline src/async_fuse/memfs/cache/memory_cache.rs:92]: I believe there's gap here. Let's assume the cache capacity is 1.
If a thread1 in here put it's own key (1,1) ,and evict the (0,0).
When thread2 in , thread2 put it's own key (2,2), and evict the (1,1), but the thread1 haven't insert (1,1) into map.
So thread2 evict nothing.
Thread1 and thread2 then insert their own key into the map.
So we can left a (1,1) in map but will never be evicted, right ?
What I want to say is , the gap between policy and map is very dangerous.
I tryied to write a test to approve it, but I failed.
Maybe something is wrong?

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+                Block::new(self.block_size)
+            })
+        };
+
+        merge_two_blocks(&block, &mut to_be_inserted);
+
+        self.write_and_evict(ino, block_id, to_be_inserted.clone())
+            .await;
+
+        self.backend
```
[the remark this answers is not in the document]
Comment (Reto911) [reply src/async_fuse/memfs/cache/memory_cache.rs:238]: We cannot simply call `load` funtion dispite of the difference of cache hit and cache miss. If we do so, we will be not able to merge the passed-in block to the cache in place, and the whole block must be copied.

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+    /// If `backend` returns a `None`, this will create a new block with zeros filled,
+    /// inserts it into the `backend`, and returns it to caller. That means, this method
+    /// will never return `None`.
+    async fn load_from_backend(&self, ino: INum, block_id: usize) -> Option<Block> {
+        if let Some(from_backend) = self.backend.load(ino, block_id).await {
+            Some(from_backend)
+        } else {
+            // If a fetching from backend still fails, create a new block filled with zeros,
+            // then write it to cache and backend as stored.
+            let mut zeroed_block = Block::new(self.block_size);
```
Comment (xiaguan) [inline src/async_fuse/memfs/cache/memory_cache.rs:186]: `new` -> `new_zeroed`

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+
+    async fn cache_block_from_backend(&self, ino: INum, block_id: usize, block: Block) {
+        self.write_block_into_cache(ino, block_id, block).await;
+    }
+
+    async fn store(&self, ino: INum, block_id: usize, mut input: IoBlock) {
+        let start_offset = input.offset();
+        let end_offset = input.end();
+
+        assert!(end_offset <= self.block_size, "out of range");
```
Comment (xiaguan) [inline src/async_fuse/memfs/cache/memory_cache.rs:210]: use `debug_assert!` instead

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+            }
+            .await;
+
+            let success = self.policy.try_put(BlockCoordinate(ino, block_id));
+
+            if success {
+                break file_cache;
+            }
+            drop(file_cache);
+            self.evict().await;
```
[the remark this answers is not in the document]
Comment (Reto911) [reply src/async_fuse/memfs/cache/memory_cache.rs:153]: Why? Could you please explain how this works?

File: src/async_fuse/memfs/cache/memory_cache.rs [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+        .get_mut(start_offset..end_offset)
+        .unwrap_or_else(|| unreachable!("`src` is checked to be in range."))
+        .copy_from_slice(src.as_slice());
+}
+
+/// The in-memory cache, implemented with lockfree hashmaps.
+#[derive(Debug)]
+pub struct InMemoryCache<P, S> {
+    /// The inner map where the cached blocks stored
+    map: HashMap<INum, Arc<RwLock<StdHashMap<usize, Block>>>>,
```
[the remark this answers is not in the document]
Comment (Reto911) [reply src/async_fuse/memfs/cache/memory_cache.rs:31]: Do you mean to define `type BlockId = usize;` ?

[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/cache/memory_cache.rs:31]: Yes

File: src/async_fuse/memfs/cache/memory_cache.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/cache/memory_cache.rs:153]: Ignore this,sry.

File: src/async_fuse/memfs/cache/memory_cache.rs (hunk 4 above)
Comment (Reto911) [reply src/async_fuse/memfs/cache/memory_cache.rs:210]: This is intentional but not for debugging, and will be replaced with an error returned in the future (the instant when error handling is introduced).

Comment (xiaguan) [reply src/async_fuse/memfs/cache/memory_cache.rs:213]: Link to the issue #425

**Review by xiaguan — APPROVED** [review]

File: src/async_fuse/memfs/cache/storage_manager.rs [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+                .unwrap_or_else(|e| panic!("Fails when awaiting on storing: {e}"));
+        }
+    }
+
+    /// Convert `IoBlock`s from slice.
+    fn make_io_blocks_from_slice(
+        &self,
+        offset: usize,
+        data: &[u8],
+    ) -> impl Iterator<Item = IoBlock> + '_ {
```
Comment (xiaguan) [inline src/async_fuse/memfs/cache/storage_manager.rs:100]: The implementation of this function seems a bit messy. I think we can refactor it. I feel like we don't need the abstraction of `IoBlock`; we can directly use Block and add other members inside it, along with corresponding initialization functions. If it's really necessary to use a data structure to represent the relationship between multiple blocks, I think `Slice` would be a better choice.

File: src/async_fuse/memfs/cache/storage_manager.rs [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+    }
+
+    /// Load data from storage.
+    pub async fn load(
+        &self,
+        ino: INum,
+        offset: usize,
+        len: usize,
+        mtime: SystemTime,
+    ) -> Vec<IoBlock> {
```
Comment (xiaguan) [inline src/async_fuse/memfs/cache/storage_manager.rs:173]: Same above.

**Review by xiaguan — APPROVED** [review]

**Pull request merged**